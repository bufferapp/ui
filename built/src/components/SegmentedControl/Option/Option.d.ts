import React from 'react';
import PropTypes from 'prop-types';
declare const Option: {
    (props: any): React.JSX.Element;
    propTypes: {
        optionType: PropTypes.Validator<string>;
        size: PropTypes.Requireable<string>;
        disabled: PropTypes.Requireable<boolean>;
        icon: PropTypes.Requireable<PropTypes.ReactNodeLike>;
        iconPosition: PropTypes.Requireable<string>;
        label: PropTypes.Validator<string>;
        value: PropTypes.Validator<string | number | boolean>;
        tooltip: PropTypes.Requireable<string>;
        selected: PropTypes.Validator<boolean>;
        onClick: PropTypes.Requireable<(...args: any[]) => any>;
    };
    defaultProps: {
        size: string;
        disabled: boolean;
        icon: null;
        iconPosition: string;
        tooltip: string;
        onClick: () => null;
    };
};
export default Option;
//# sourceMappingURL=Option.d.ts.map