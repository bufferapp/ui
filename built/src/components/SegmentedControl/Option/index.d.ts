export { default } from './Option';
//# sourceMappingURL=index.d.ts.map