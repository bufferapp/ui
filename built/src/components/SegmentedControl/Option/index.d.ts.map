{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["../../../../../src/components/SegmentedControl/Option/index.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,OAAO,EAAE,MAAM,UAAU,CAAA"}