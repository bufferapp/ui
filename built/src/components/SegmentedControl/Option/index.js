export { default } from './Option';
//# sourceMappingURL=index.js.map