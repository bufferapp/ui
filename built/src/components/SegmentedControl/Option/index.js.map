{"version":3,"file":"index.js","sourceRoot":"","sources":["../../../../../src/components/SegmentedControl/Option/index.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,OAAO,EAAE,MAAM,UAAU,CAAA","sourcesContent":["export { default } from './Option'\n"]}