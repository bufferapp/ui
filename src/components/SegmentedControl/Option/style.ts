import styled, { css } from 'styled-components'
import {
  blue,
  blueLightest,
  boxShadow,
  gray,
  grayDark,
  grayLight,
  transparent,
} from '../../style/colors'
import { borderRadius } from '../../style/borders'
import { fontSize, fontWeightMedium } from '../../style/fonts'

// @ts-expect-error TS(7006) FIXME: Parameter 'iconPosition' implicitly has an 'any' t... Remove this comment to see the full error message
const getGridTemplateAreas = (iconPosition) => {
  switch (iconPosition) {
    case 'top':
      return `
        "icon"
        "label"
      `

    case 'bottom':
      return `
        "label"
        "icon"
      `

    case 'right':
      return `"label icon"`

    case 'left':
    default:
      return `"icon label"`
  }
}

export const OptionStyled = styled.div`
  display: flex;
  align-items: center;
  cursor: ${(props) => (props.disabled ? 'not-allowed' : 'pointer')};
  border-radius: ${borderRadius};
  margin-right: 4px;
  transition: all 0.1s ease-out;

  border: ${(props) => {
    if (props.selected) return `1px solid ${blueLightest}`
    if (props.selected) return `1px solid ${gray}`
    return `1px solid ${transparent}`
  }};

  color: ${(props) => {
    if (props.selected) return blue
    if (props.disabled) return gray
    return grayDark
  }};

  background-color: ${(props) => {
    if (props.selected) return blueLightest
    if (props.disabled) return grayLight
  }};

  padding: ${(props) => {
    if (props.size === 'large') return '11px'
    if (props.size === 'small') return '3px 7px'
    return '7px'
  }};

  &:last-child {
    margin-right: 0;
  }

  ${(props) =>
    !props.disabled &&
    css`
      :hover {
        color: ${blue};
      }

      :active {
        background-color: ${blueLightest};
      }

      :focus {
        outline: none;
        border: 1px solid ${blue};
        box-shadow: 0 0 0 1px ${boxShadow};
      }
    `}
`

export const ContentWrapper = styled.div`
  display: grid;
  place-items: center;
  // Calculate grid-template-rows value based on provided iconPosition
  grid-template-areas: ${(props) => getGridTemplateAreas(props.iconPosition)};
  // if horizontal icon position, two columns one row
  grid-template-columns: ${(props) =>
    ['left', 'right'].includes(props.iconPosition) ? 'auto auto' : 'auto'};
  // if vertical icon position, one column two rows
  grid-template-rows: ${(props) =>
    ['top', 'bottom'].includes(props.iconPosition) ? 'auto' : 'auto auto'};
`

export const IconWrapper = styled.div`
  display: flex;
  grid-area: icon;
`

export const Label = styled.span`
  font-weight: ${fontWeightMedium};
  font-size: ${fontSize};
  grid-area: label;
  ${(props) =>
    props.optionType === 'textAndIcon' && `margin-${props.iconPosition}: 10px`};
`
