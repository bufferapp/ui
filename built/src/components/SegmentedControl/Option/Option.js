import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import Tooltip from '../../Tooltip';
import { OptionStyled, ContentWrapper, Label, IconWrapper } from './style';
const Option = (props) => {
    const { selected, disabled, icon, label, value, optionType, iconPosition, size, tooltip, onClick, } = props;
    const showIcon = optionType === 'icon' || optionType === 'textAndIcon';
    const showLabel = optionType === 'text' || optionType === 'textAndIcon';
    const tooltipMessage = useMemo(() => {
        if (tooltip)
            return tooltip;
        if (optionType === 'icon')
            return `View ${label} only`;
    }, [optionType, tooltip, label]);
    const handleClick = () => {
        if (!disabled)
            onClick(value);
    };
    const renderContent = () => (React.createElement(ContentWrapper, { iconPosition: iconPosition },
        icon && showIcon && React.createElement(IconWrapper, null, icon),
        label && showLabel && (React.createElement(Label, { optionType: optionType, iconPosition: iconPosition }, label))));
    return (React.createElement(OptionStyled, { selected: !disabled && selected, "aria-pressed": !disabled && selected, disabled: disabled, size: size, onClick: handleClick, tabIndex: "0", role: "button" }, tooltipMessage ? (React.createElement(Tooltip, { label: tooltipMessage }, renderContent())) : (renderContent())));
};
Option.propTypes = {
    optionType: PropTypes.oneOf(['text', 'icon', 'textAndIcon']).isRequired,
    size: PropTypes.oneOf(['small', 'normal', 'large']),
    disabled: PropTypes.bool,
    icon: PropTypes.node,
    iconPosition: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
    label: PropTypes.string.isRequired,
    value: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
        PropTypes.bool,
    ]).isRequired,
    tooltip: PropTypes.string,
    selected: PropTypes.bool.isRequired,
    onClick: PropTypes.func,
};
Option.defaultProps = {
    size: 'normal',
    disabled: false,
    icon: null,
    iconPosition: 'left',
    tooltip: '',
    onClick: () => null,
};
export default Option;
//# sourceMappingURL=Option.js.map