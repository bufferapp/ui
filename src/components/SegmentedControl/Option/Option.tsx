import React, { useMemo } from 'react'
import PropTypes from 'prop-types'

import Tooltip from '../../Tooltip'
import { OptionStyled, ContentWrapper, Label, IconWrapper } from './style'

/**
 * Option component for Segmented Control component.
 */
// @ts-expect-error TS(7006) FIXME: Parameter 'props' implicitly has an 'any' type.
const Option = (props) => {
  const {
    selected,
    disabled,
    icon,
    label,
    value,
    optionType,
    iconPosition,
    size,
    tooltip,
    onClick,
  } = props

  // Whether or not to attempt to render an icon
  const showIcon = optionType === 'icon' || optionType === 'textAndIcon'
  // Whether or not to render a label
  const showLabel = optionType === 'text' || optionType === 'textAndIcon'

  const tooltipMessage = useMemo(() => {
    // If tooltip message is provided, show it
    if (tooltip) return tooltip
    // If no tooltip message is provided, but optionType is 'icon'
    // derive a tooltip message from the provided label
    if (optionType === 'icon') return `View ${label} only`
  }, [optionType, tooltip, label])

  const handleClick = () => {
    // Only run onClick callback if option is not disabled
    if (!disabled) onClick(value)
  }

  const renderContent = () => (
    // @ts-expect-error TS(2769) FIXME: No overload matches this call.
    <ContentWrapper iconPosition={iconPosition}>
      {icon && showIcon && <IconWrapper>{icon}</IconWrapper>}

      {label && showLabel && (
        // @ts-expect-error TS(2769) FIXME: No overload matches this call.
        <Label optionType={optionType} iconPosition={iconPosition}>
          {label}
        </Label>
      )}
    </ContentWrapper>
  )

  return (
    <OptionStyled
      selected={!disabled && selected}
      aria-pressed={!disabled && selected}
      disabled={disabled}
      size={size}
      onClick={handleClick}
      // @ts-expect-error TS(2769) FIXME: No overload matches this call.
      tabIndex="0"
      role="button"
    >
      {tooltipMessage ? (
        // @ts-expect-error TS(2322) FIXME: Type '{ children: Element; label: any; }' is not a... Remove this comment to see the full error message
        <Tooltip label={tooltipMessage}>{renderContent()}</Tooltip>
      ) : (
        renderContent()
      )}
    </OptionStyled>
  )
}

Option.propTypes = {
  /** Type of options. Options: text, icon, textAndIcon. */
  optionType: PropTypes.oneOf(['text', 'icon', 'textAndIcon']).isRequired,

  /** Size of control. Options: small, normal, large. */
  size: PropTypes.oneOf(['small', 'normal', 'large']),

  /** It disables the option. </i> */
  disabled: PropTypes.bool,

  /** The title of the icon to pass to the Icon component */
  icon: PropTypes.node,

  /** The position of the icon relative to the label. Can be `top`, `right`, `bottom`, `left`. */
  iconPosition: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),

  /** The option label */
  label: PropTypes.string.isRequired,

  /** The option value */
  value: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.bool,
  ]).isRequired,

  /** Optional custom tooltip message to display on hover */
  tooltip: PropTypes.string,

  /** Whether the option is the selected one */
  selected: PropTypes.bool.isRequired,

  /** onClick callback function */
  onClick: PropTypes.func,
}

Option.defaultProps = {
  size: 'normal',
  disabled: false,
  icon: null,
  iconPosition: 'left',
  tooltip: '',
  onClick: () => null,
}

export default Option
