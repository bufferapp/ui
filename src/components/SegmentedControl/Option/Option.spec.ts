// @ts-expect-error TS(7016) FIXME: Could not find a declaration file for module 'jest... Remove this comment to see the full error message
import snap from 'jest-auto-snapshots'
import Option from './Option'

snap(Option, './Option.jsx')
