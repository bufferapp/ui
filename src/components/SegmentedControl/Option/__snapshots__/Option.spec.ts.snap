// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-auto-snapshots > Option Matches snapshot when boolean prop "disabled" is set to: "false" 1`] = `
<styled.div
  aria-pressed={true}
  disabled={false}
  onClick={[Function]}
  role="button"
  selected={true}
  size="small"
  tabIndex="0"
>
  <Tooltip
    customLabel=""
    hotkey=""
    label="jest-auto-snapshots String Fixture"
    opacity={1}
    position="bottom"
    verticalAlign="center"
  >
    <styled.div
      iconPosition="top"
    >
      <styled.span
        iconPosition="top"
        optionType="text"
      >
        jest-auto-snapshots String Fixture
      </styled.span>
    </styled.div>
  </Tooltip>
</styled.div>
`;

exports[`jest-auto-snapshots > Option Matches snapshot when boolean prop "selected" is set to: "false" 1`] = `
<styled.div
  aria-pressed={false}
  disabled={true}
  onClick={[Function]}
  role="button"
  selected={false}
  size="small"
  tabIndex="0"
>
  <Tooltip
    customLabel=""
    hotkey=""
    label="jest-auto-snapshots String Fixture"
    opacity={1}
    position="bottom"
    verticalAlign="center"
  >
    <styled.div
      iconPosition="top"
    >
      <styled.span
        iconPosition="top"
        optionType="text"
      >
        jest-auto-snapshots String Fixture
      </styled.span>
    </styled.div>
  </Tooltip>
</styled.div>
`;

exports[`jest-auto-snapshots > Option Matches snapshot when passed all props 1`] = `
<styled.div
  aria-pressed={false}
  disabled={true}
  onClick={[Function]}
  role="button"
  selected={false}
  size="small"
  tabIndex="0"
>
  <Tooltip
    customLabel=""
    hotkey=""
    label="jest-auto-snapshots String Fixture"
    opacity={1}
    position="bottom"
    verticalAlign="center"
  >
    <styled.div
      iconPosition="top"
    >
      <styled.span
        iconPosition="top"
        optionType="text"
      >
        jest-auto-snapshots String Fixture
      </styled.span>
    </styled.div>
  </Tooltip>
</styled.div>
`;

exports[`jest-auto-snapshots > Option Matches snapshot when passed only required props 1`] = `
<styled.div
  aria-pressed={true}
  disabled={false}
  onClick={[Function]}
  role="button"
  selected={true}
  size="normal"
  tabIndex="0"
>
  <styled.div
    iconPosition="left"
  >
    <styled.span
      iconPosition="left"
      optionType="text"
    >
      jest-auto-snapshots String Fixture
    </styled.span>
  </styled.div>
</styled.div>
`;
